### ISSUE TITLE
sdists are missing long_description

### ISSUE THREAD

**Issue by merwok** [takluyver/flit#216]
To switch a project to flit, I changed my `setup.py check` command with readme_renderer to `twine check`.  I get this traceback:

```
Checking distribution .tox/dist/Rest_Framework_Auth_Toolkit-0.9.dev0.tar.gz: ** None  {}
Traceback (most recent call last):
  File "project/.tox/pkg/bin/twine", line 11, in <module>
    sys.exit(main())
  File "project/.tox/pkg/lib/python3.6/site-packages/twine/__main__.py", line 28, in main
    return dispatch(sys.argv[1:])
  File "project/.tox/pkg/lib/python3.6/site-packages/twine/cli.py", line 75, in dispatch
    return main(args.args)
  File "project/.tox/pkg/lib/python3.6/site-packages/twine/commands/check.py", line 127, in main
    return check(args.dists)
  File "project/.tox/pkg/lib/python3.6/site-packages/twine/commands/check.py", line 98, in check
    metadata.get("description"), stream=stream, **parameters
  File "project/.tox/pkg/lib/python3.6/site-packages/readme_renderer/rst.py", line 95, in render
    parts = publish_parts(raw, writer=writer, settings_overrides=settings)
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/core.py", line 450, in publish_parts
    enable_exit_status=enable_exit_status)
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/core.py", line 664, in publish_programmatically
    output = pub.publish(enable_exit_status=enable_exit_status)
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/core.py", line 217, in publish
    self.settings)
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/readers/__init__.py", line 71, in read
    self.input = self.source.read()
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/io.py", line 435, in read
    return self.decode(self.source)
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/io.py", line 99, in decode
    data_encoding = self.determine_encoding_from_data(data)
  File "project/.tox/pkg/lib/python3.6/site-packages/docutils/io.py", line 142, in determine_encoding_from_data
    if data.startswith(start_bytes):
AttributeError: 'NoneType' object has no attribute 'startswith'
```

The twine bug is https://github.com/pypa/twine/issues/411
The flit bug is that generated setup.py does not contain long_description.

(And I do want to check both the wheel and the sdist; even if PyPI prefers the wheel or the most recently updated dist or something else, I’d like to know that each artifact has correct contents.)

**Comment by di** [comment]
IMO flit is behaving correctly here, this is simply a bug in twine.

**Comment by merwok** [comment]
I am!

**Comment by di** [comment]
You're right, here's a minimal example with `flit==1.2.1`:

```
$ cat pyproject.toml
[build-system]
requires = ["flit"]
build-backend = "flit.buildapi"

[tool.flit.metadata]
module = "blah"
author = "blah"
author-email = "blah@blah.blah"
home-page = "http://blah.blah"
classifiers = ["License :: OSI Approved :: Apache Software License"]
description-file = "README.md"
```

```
$ unzip -p dist/blah-0.1-py2.py3-none-any.whl blah-0.1.dist-info/METADATA
Metadata-Version: 2.1
Name: blah
Version: 0.1
Summary: An amazing sample package!
Home-page: http://blah.blah
License: UNKNOWN
Author: blah
Author-email: blah@blah.blah
Description-Content-Type: text/markdown
Classifier: License :: OSI Approved :: Apache Software License

Blah blah blah
```

```
$ tar xfO dist/blah-0.1.tar.gz blah-0.1/PKG-INFO
Metadata-Version: 1.1
Name: blah
Version: 0.1
Summary: An amazing sample package!
Home-page: http://blah.blah
Author: blah
Author-email: blah@blah.blah
```

**Comment by takluyver** [comment]
BTW, Flit already checks the README and shows a warning if it's invalid rst. We should probably add an option to error out in that case.

**Comment by merwok** [comment]
Does flit check if the markdown or reST is well formed?

**Comment by di** [comment]
FWIW there's no such thing as "well formed Markdown", it will fail gracefully if syntax is wrong, unlike rST, thus it will always render as Markdown.

**Comment by sethmlarson** [comment]
Confirming that this issue still exists, perhaps I can work on a fix as well.

**Comment by merwok** [comment]
PKG-INFO files seem fully complete after switching to PEP 621 metadata.

**Comment by takluyver** [comment]
Great, it was probably #410 that fixed this (released in Flit 3.3).
